cmd/ci-operator-configresolver/main.go

```diff
@@ -155,30 +156,49 @@ func main() {
 	opts := options{}
 	flag.StringVar(&opts.regPath, "registry", "", "Path to step registry")
 	flag.StringVar(&opts.configPath, "config", "", "Path to config dirs")
+	verbose := flag.Bool("verbose", false, "Enable debug logging messages")
 	flag.Parse()
 	reloadRegistry(&opts)
+	if *verbose {
+		log.SetLevel(log.DebugLevel)
+	}
 
 	// watch for changes in config directories
 	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		log.Fatalf("Failed to create new watcher: %v", err)
 	}
 	defer watcher.Close()
+	reloadSet := false
 	go func() {
 		for {
 			select {
-			case _ = <-watcher.Events:
-				log.Info("Reloading registry")
-				if err := reloadRegistry(&opts); err != nil {
-					log.Fatalf("Failed to reload registry: %v", err)
+			case event := <-watcher.Events:
+				log.Debugf("Received %v event for %s", event.Op, event.Name)
+				if event.Op == fsnotify.Remove {
+					log.Debugf("Removing %s from watches", event.Name)
+					watcher.Remove(event.Name)
+				}
+				if !reloadSet {
+					reloadSet = true
+					// use separate go routine to allow other events to run while the reloader waits
+					go func() {
+						time.Sleep(time.Second)
+						// unset reloadSet before running reload function in case a change occurs during the reload
+						reloadSet = false
```

Comment:
You're writing and reading a value from separate routines, which will be racy. Let's use the `sync` package for concurrency: (this is rough, will need refinement and testing)

```go
type coalescer struct {
    lock sync.Mutex
    loading bool
    loader sync.Cond
    latest *Registry
}

func (c *coalescer) reload() {
    c.Lock()
    if c.loading {
        // someone else is loading it, so we wait
        c.loader.L.Lock()
        c.Unlock()
        c.loader.Wait()
        c.loader.L.Unlock()
        return c.latest
    }
    // nobody else is loading
    c.loading = true
    c.Unlock()

    // reload the registry, wait an extra second after if you want
    registry := registry.Load()
    c.Lock()
    c.latest = registry
    c.Unlock()

    // inform the waiters that we are done
    c.loader.L.Lock()
    c.loader.Broadcast()
    c.loader.L.Unlock()
}
```